### PR TITLE
Fix bugs in `VisitorPartialEvaluator`, namely with support for `char` and preservation of types

### LINKED ISSUE #5273: [Bug]: `CtCodeElement#partiallyEvaluate` crashes for `c < 's'` `java.lang.Character cannot be cast to class java.lang.Number`
### Describe the bug

`VisitorPartialEvaluator#visitCtBinaryOperator` unconditionally casts the literal of a `CtBinaryOperator` to `Number` for certain operators. `Character` does not implement `Number`, so the code crashes.

### Source code you are trying to analyze/transform

```Java
public class Main {
    public static boolean evaluate(char c) {
        return c < 'a';
    }
}
```

### Source code for your Spoon processing

```Java
import spoon.Launcher;
import spoon.reflect.CtModel;
import spoon.reflect.code.CtBinaryOperator;
import spoon.reflect.visitor.filter.TypeFilter;
import spoon.support.compiler.VirtualFile;

public final class Main2 {
    private Main2() {
    }

    private static void run() {
        Launcher launcher = new Launcher();

        launcher.getEnvironment().setComplianceLevel(17);
        launcher.addInputResource(new VirtualFile("""
            public class Main {
                public static boolean evaluate(char c) {
                    return c < 'a';
                }
            }
            """));

        CtModel ctModel = launcher.buildModel();

        CtBinaryOperator<?> ctBinaryOperator = ctModel.getElements(new TypeFilter<>(CtBinaryOperator.class)).get(0);
        System.out.println(ctBinaryOperator.partiallyEvaluate());
    }

    public static void main(String[] args) {
        run();
    }
}
```

### Actual output

```Java
Exception in thread "main" java.lang.ClassCastException: class spoon.support.reflect.code.CtBinaryOperatorImpl cannot be cast to class java.lang.String (spoon.support.reflect.code.CtBinaryOperatorImpl is in unnamed module of loader 'app'; java.lang.String is in module java.base of loader 'bootstrap')
	at org.example.playground.Main2.run(Main2.java:64)
	at org.example.playground.Main2.main(Main2.java:68)
```

### Expected output

```Java
c < 'a'
```

### Spoon Version

10.3.0

### JVM Version

openjdk version "17.0.1" 2021-10-19 OpenJDK Runtime Environment Temurin-17.0.1+12 (build 17.0.1+12) OpenJDK 64-Bit Server VM Temurin-17.0.1+12 (build 17.0.1+12, mixed mode, sharing)

### What operating system are you using?

Windows 10

### LINKED ISSUE #5289: [Bug]: `VisitorPartialEvaluator` creates `CtLiteral`s without a set `CtType`
### Describe the bug

The `CtLiteral`s emitted by the `VisitorPartialEvaluator` do not have a set type. This causes problems, because `CtLiteral#getFactory()` returns `null`.

### Source code you are trying to analyze/transform

```Java
public class Main {
    public static int evaluate() {
        return 1 + 2;
    }
}
```

### Source code for your Spoon processing

```Java
import spoon.Launcher;
import spoon.reflect.CtModel;
import spoon.reflect.code.CtBinaryOperator;
import spoon.reflect.code.CtLiteral;
import spoon.reflect.visitor.filter.TypeFilter;
import spoon.support.compiler.VirtualFile;

public final class Main2 {
    private Main2() {
    }

    private static void run() {
        Launcher launcher = new Launcher();

        launcher.getEnvironment().setComplianceLevel(17);
        launcher.addInputResource(new VirtualFile("""
            public class Main {
                public static int evaluate() {
                    return 1 + 2;
                }
            }
            """));

        CtModel ctModel = launcher.buildModel();

        CtBinaryOperator<?> ctBinaryOperator = ctModel.getElements(new TypeFilter<>(CtBinaryOperator.class)).get(0);
        CtLiteral<Integer> result = ctBinaryOperator.partiallyEvaluate();
        System.out.println("base: %s\nvalue: %s\ntype: %s\nvalue class: %s".formatted(
            result.getBase(),
            result.getValue(),
            result.getType(),
            result.getValue().getClass()
        ));
    }

    public static void main(String[] args) {
        run();
    }
}
```

### Actual output

```Java
base: null
value: 3
type: null
value class: class java.lang.Integer
```

### Expected output

```Java
base: null
value: 3
type: int
value class: class java.lang.Integer
```

### Spoon Version

10.3.0

### JVM Version

openjdk version "17.0.1" 2021-10-19 OpenJDK Runtime Environment Temurin-17.0.1+12 (build 17.0.1+12) OpenJDK 64-Bit Server VM Temurin-17.0.1+12 (build 17.0.1+12, mixed mode, sharing)

### What operating system are you using?

Windows 10

### PR DESCRIPTION
**Pull request by Luro02**
There are multiple issues with `VisitorPartialEvaluator`:
- [x] should not crash with `char`-literals #5273
- [x] the type should be set for emitted `CtLiterals` #5289 
- [x] should not crash with unsupported `CtUnaryOperator` (instead yield the original code without simplifying it)
- [x] the `isLiteralType` method should return `true` for a `boolean` literal
- [x] type casts should not be ignored in evaluation (otherwise code like `(byte) 400 + 20` will not evaluate correctly), bug found by @I-Al-Istannen 

(and more that I do not want to fix, like a better way to set the result, making `getResult` and `setResult` protected for inheritance, ensuring everything else has a correct type, ...)

This PR...
- closes #5273 
- closes #5289

### CODE REVIEW

File: src/main/java/spoon/reflect/visitor/OperatorHelper.java [hunk 1]
```diff
@@ -219,4 +244,277 @@ public static OperatorAssociativity getOperatorAssociativity(UnaryOperatorKind o
...
+				if (operand.getType().unbox().equals(typeFactory.BOOLEAN_PRIMITIVE)) {
+					return Optional.of(typeFactory.BOOLEAN_PRIMITIVE);
+				}
+
+				return Optional.empty();
+			case PREINC:
+			case PREDEC:
+			case POSTINC:
+			case POSTDEC:
+				// Not yet implemented, because it is not necessary for the current use case.
```
Comment (SirYwell) [inline src/main/java/spoon/reflect/visitor/OperatorHelper.java:514]: This should be either addressed or documented in an issue so we can keep track of it.

File: src/main/java/spoon/reflect/visitor/OperatorHelper.java [hunk 2]
```diff
@@ -219,4 +244,277 @@ public static OperatorAssociativity getOperatorAssociativity(UnaryOperatorKind o
...
+					return Optional.empty();
+				}
+
+				// The type of the shift expression is the promoted type of the left-hand operand.
+				return Optional.of(promotedLeft);
+			}
+			case INSTANCEOF:
+				// See: https://docs.oracle.com/javase/specs/jls/se11/html/jls-15.html#jls-15.20.2
+				// Not implemented, because it is not necessary for the current use case.
+				throw new IllegalStateException("instanceof is not yet implemented");
```
Comment (SirYwell) [inline src/main/java/spoon/reflect/visitor/OperatorHelper.java:379]: Same as with EQ, NE; I think this can just always be boolean?

**Review by SirYwell — CHANGES_REQUESTED** [review]
I hope I got most of the things with the first review, this really contains a lot of edge cases to consider.

I'm not sure about some checks that are only needed when considering invalid code, but that's fine probably.

Generally, the code and the tests look good already, but there are a few places that need further attention/adjustments.

File: src/test/java/spoon/test/eval/EvalTest.java [hunk 3]
```diff
@@ -343,4 +396,182 @@ void testDoublePrecisionLost(String literal, String expected) {
...
+		Map.entry(short.class, factory -> factory.createLiteral((short) 1)),
+		Map.entry(int.class, factory -> factory.createLiteral((int) 1)),
+		Map.entry(long.class, factory -> factory.createLiteral(1L)),
+		Map.entry(float.class, factory -> factory.createLiteral(1.0f)),
+		Map.entry(double.class, factory -> factory.createLiteral(1.0d)),
+		Map.entry(boolean.class, factory -> factory.createLiteral(true)),
+		Map.entry(char.class, factory -> factory.createLiteral('a')),
+		Map.entry(String.class, factory -> factory.createLiteral("a")),
+		// not sure what type to use for null, so use Optional
+		Map.entry(Optional.class, factory -> factory.createLiteral(null))
```
[the remark this answers is not in the document]
Comment (Luro02) [reply src/test/java/spoon/test/eval/EvalTest.java:411]: That was my initial thought as well, but IntelliJ complained about duplicate keys. I am not sure how hashCode and equals works on `Class<T>`, so I will look into this

File: src/main/java/spoon/reflect/visitor/OperatorHelper.java (hunk 2 above)
Comment (Luro02) [reply src/main/java/spoon/reflect/visitor/OperatorHelper.java:379]: It seems like I have to document better what this method does. The method does **not** return the type of the expression (that would be unnecessary, because `CtBinaryOperator#getType` will return it).

Instead it will check if an expression can be applied to an operator (if not, it will return empty).

There are some types like `short` that are cast to another type (in this case `int`) before the operator is applied. The types seem to always be the same after promotion on both sides, so there is no distinction made.

Returning `boolean` for the instanceof operator, would result in the left and right being cast to `boolean` before applying the operator (which is obviously wrong).

Thinking about 
> types seem to always be the same after promotion on both sides

it seems like that would not apply to `instanceof`.

File: src/main/java/spoon/reflect/visitor/OperatorHelper.java [hunk 4]
```diff
@@ -219,4 +244,277 @@ public static OperatorAssociativity getOperatorAssociativity(UnaryOperatorKind o
...
+						if (leftType.equals(nullType)) {
+							return Optional.of(rightType);
+						}
+
+						if (rightType.equals(nullType)) {
+							return Optional.of(leftType);
+						}
+
+						Class<?> leftClass = leftType.getActualClass();
+						Class<?> rightClass = rightType.getActualClass();
```
[the remark this answers is not in the document]
Comment (Luro02) [reply src/main/java/spoon/reflect/visitor/OperatorHelper.java:414]: I had not found a method in spoon that replicates the `isAssignableFrom` method, which does exactly what the specification requires.

I did not want to spend time implementing even more of the specification, which is why I settled for this. `VisitorPartialEvaluator` only uses this method with `CtLiteral`s, so the possible types are limited to class literals, strings, and primitive types. I think for all of them `getActualClass` should work.

File: src/main/java/spoon/support/reflect/eval/VisitorPartialEvaluator.java [hunk 5]
```diff
@@ -115,96 +117,161 @@ void setResult(CtElement element) {
...
 			case MUL:
-				res.setValue(convert(operator.getType(),
-								((Number) leftObject).doubleValue() * ((Number) rightObject).doubleValue()));
+				value = convert(operator.getType(),
+					((Number) leftObject).doubleValue() * ((Number) rightObject).doubleValue());
 				break;
 			case DIV:
-				res.setValue(convert(operator.getType(),
-								((Number) leftObject).doubleValue() / ((Number) rightObject).doubleValue()));
+				value = convert(operator.getType(),
```
[the remark this answers is not in the document]
Comment (Luro02) [reply src/main/java/spoon/support/reflect/eval/VisitorPartialEvaluator.java]: Was not much more effort to implement, so I did implement it. Not sure about the try catch, I could instead check if the divisor is zero beforehand.

File: src/main/java/spoon/reflect/visitor/OperatorHelper.java (hunk 2 above)
Comment (Luro02) [reply src/main/java/spoon/reflect/visitor/OperatorHelper.java:392]: The problem is that I have to check any context where `lhs instanceof rhs` is not a valid operation. According to the [JLS](https://docs.oracle.com/javase/specs/jls/se11/html/jls-15.html#jls-15.20.2), I would have to check
- that the lhs is a reference type or null type (should be easy to implement)
- that the rhs is [reifiable](https://docs.oracle.com/javase/specs/jls/se11/html/jls-4.html#jls-4.7) (sounds like a lot more work)

File: src/main/java/spoon/reflect/visitor/OperatorHelper.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (Luro02) [reply src/main/java/spoon/reflect/visitor/OperatorHelper.java:427]: The promoted type is not `boolean`, the subclass would be cast to it's parent class. (So the type would be the one the other side is assignable to)

[the remark this answers is not in the document]
Comment (I-Al-Istannen) [reply src/main/java/spoon/reflect/visitor/OperatorHelper.java:427]: > I had not found a method in spoon that replicates the isAssignableFrom method, which does exactly what the specification requires.

Isn't this just a subtype check (e.g. using `TypeAdaptor#isSubtypeOf`)?

**Review by SirYwell — APPROVED** [review]
Looks good.

**Comment by SirYwell** [comment]
Thanks @Luro02
